File: example/javalib/web/1-web-example/test/src/WebServerTests.java
package example;

import static org.junit.jupiter.api.Assertions.*;

import okhttp3.*;
import org.junit.jupiter.api.*;
import org.springframework.boot.SpringApplication;
import org.springframework.context.ConfigurableApplicationContext;

public class WebServerTests {
  private static ConfigurableApplicationContext server;
  private static final OkHttpClient client = new OkHttpClient();

  @BeforeAll
  public static void startServer() {
    server = SpringApplication.run(WebServer.class, new String[] {});
  }

  @AfterAll
  public static void stopServer() {
    if (server != null) server.close();
  }

  @Test
  public void testReverseString() throws Exception {
    var body = RequestBody.create("helloworld", MediaType.parse("text/plain"));

    var request = new Request.Builder()
        .url("http://localhost:8080/reverse-string")
        .post(body)
        .build();

    try (var response = client.newCall(request).execute()) {
      assertTrue(response.isSuccessful());
      assertEquals("dlrowolleh", response.body().string());
    }
  }
}

File: example/kotlinlib/config/7-web-server/test/src/WebServerTests.kt
package example

import io.ktor.client.*
import io.ktor.client.engine.cio.CIO as ClientCIO
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.server.engine.*
import io.ktor.server.cio.CIO as ServerCIO
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*
import kotlinx.coroutines.runBlocking

class WebServerTests {
    companion object {
        private lateinit var server: ApplicationEngine
        private val client = HttpClient(ClientCIO)

        @BeforeAll
        @JvmStatic
        fun startServer() {
            server = embeddedServer(ServerCIO, port = 8080) {
                routing {
                    post("/reverse-string") {
                        val body = call.receiveText()
                        call.respondText(body.reversed())
                    }
                }
            }
            server.start(wait = false)
            Thread.sleep(1000) // Give server time to start
        }

        @AfterAll
        @JvmStatic
        fun stopServer() {
            server.stop(1000, 2000)
            client.close()
        }
    }

    @Test
    fun testReverseString() = runBlocking {
        val response = client.post("http://localhost:8080/reverse-string") {
            contentType(ContentType.Text.Plain)
            setBody("helloworld")
        }
        assertEquals(200, response.status.value)
        assertEquals("dlrowolleh", response.bodyAsText())
    }
}
